File: leetcode_problems/python/lc427_quad_tree.py
from typing import List
from pprint import pprint


class Node:
    def __init__(
        self,
        val: int,
        isLeaf: bool,
        topLeft=None,
        topRight=None,
        bottomLeft=None,
        bottomRight=None,
    ):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight


class Solution:
    def create(
        self,
        grid: List[List[int]],
        cmin: int,
        cmax: int,
        rmin: int,
        rmax: int,
    ) -> Node:
        cspread = cmax - cmin
        rspread = rmax - rmin
        if cspread <= 1 or rspread <= 1:
            return Node(grid[cmin][rmin], True, None, None, None, None)

        cmid = cmin + (cmax - cmin) // 2
        rmid = rmin + (rmax - rmin) // 2

        tl = self.create(grid, cmin, cmid, rmin, rmid)
        tr = self.create(grid, cmin, cmid, rmid, rmax)
        bl = self.create(grid, cmid, cmax, rmin, rmid)
        br = self.create(grid, cmid, cmax, rmid, rmax)

        check1 = tl.isLeaf == tr.isLeaf == bl.isLeaf == br.isLeaf == True
        check2 = tl.val == tr.val == bl.val == br.val
        if check1 and check2:
            node = Node(tl.val, True, None, None, None, None)

        else:
            node = Node(0, False, tl, tr, bl, br)

        return node

    def construct(
        self,
        grid: List[List[int]],
    ) -> "Node":
        return self.create(
            grid,
            cmin=0,
            cmax=len(grid),
            rmin=0,
            rmax=len(grid),
        )


if __name__ == "__main__":
    sol = Solution()

    grid = [
        [1, 1, 1, 1, 0, 0, 0, 0],
        [1, 1, 1, 1, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 0, 0, 0, 0],
        [1, 1, 1, 1, 0, 0, 0, 0],
        [1, 1, 1, 1, 0, 0, 0, 0],
        [1, 1, 1, 1, 0, 0, 0, 0],
    ]

    quad_tree = sol.construct(grid)

File: leetcode_problems/typescript/src/lc427_quad_tree.ts
class _Node {
  val: boolean;
  isLeaf: boolean;
  topLeft: _Node | null;
  topRight: _Node | null;
  bottomLeft: _Node | null;
  bottomRight: _Node | null;
  constructor(
    val?: boolean,
    isLeaf?: boolean,
    topLeft?: _Node,
    topRight?: _Node,
    bottomLeft?: _Node,
    bottomRight?: _Node
  ) {
    this.val = val === undefined ? false : val;
    this.isLeaf = isLeaf === undefined ? false : isLeaf;
    this.topLeft = topLeft === undefined ? null : topLeft;
    this.topRight = topRight === undefined ? null : topRight;
    this.bottomLeft = bottomLeft === undefined ? null : bottomLeft;
    this.bottomRight = bottomRight === undefined ? null : bottomRight;
  }
}

function createQuad(
  grid: number[][],
  cmin: number,
  cmax: number,
  rmin: number,
  rmax: number
): _Node {
  let cspread = cmax - cmin;
  let rspread = rmax - rmin;

  if (cspread <= 1 || rspread <= 1) {
    return new _Node(Boolean(grid[cmin][rmin]), true);
  }

  let cmid = cmin + Math.floor((cmax - cmin) / 2);
  let rmid = rmin + Math.floor((rmax - rmin) / 2);

  let tl = createQuad(grid, cmin, cmid, rmin, rmid);
  let tr = createQuad(grid, cmin, cmid, rmid, rmax);
  let bl = createQuad(grid, cmid, cmax, rmin, rmid);
  let br = createQuad(grid, cmid, cmax, rmid, rmax);

  let check1 = tl.isLeaf && tr.isLeaf && bl.isLeaf && br.isLeaf && true;
  let check2 = tl.val === tr.val && tl.val === bl.val && tl.val === br.val;

  if (check1 && check2) {
    let node = new _Node(tl.val, true);
    return node;
  } else {
    let node = new _Node(false, false, tl, tr, bl, br);
    return node;
  }
}

function constructQuad(grid: number[][]): _Node | null {
  return createQuad(grid, 0, grid.length, 0, grid.length);
}
